icepack/statistics.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (C) 2022 by Daniel Shapero <shapero@uw.edu>
+#
+# This file is part of icepack.
+#
+# icepack is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# The full text of the license can be found in the file LICENSE in the
+# icepack source directory or at <http://www.gnu.org/licenses/>.
+
+from copy import deepcopy
+import firedrake
+import pyadjoint, pyadjoint.optimization.rol_solver
+
+
+class StatisticsProblem:
+    def __init__(self, simulation, likelihood, prior, controls):
+        r"""Initialize a data assimilation problem
+
+        Parameters
+        ----------
+        simulation
+            A function that takes in the controls and returns the observables
+        likelihood
+            A function that takes in the observables returns a form that can be
+            assembled to produce the value of the model-data misfit, i.e. the
+            negative logarithm of the likelihood function
+        prior
+            A function that takes in the controls and returns a form that can
+            be assembled to produce the value of the regularization, i.e. the
+            negative log prior
```

Comment:
I appreciate this nod to bayesian thinking but is this the standard way of doing it? Them _actually_ being the the negative logarithms confused me when I first read this. Could they be called something like `neg_log_likelihood` and `neg_log_prior`? If it's the convention then I will bow to it though

Reply:
Given that negative log likelihood is an actual stats term perhaps `neg_log_likelihood` or `negative_log_likelihood`? Log likelihood is also apparently often just a lower case l so I suppose `neg_l` or `negative_l` (though this is less self explanatory).

Negative log prior is not something I seem to be able to find much information about (at least on page 1 of google). I guess it depends whether your prior always ought to be considered with respect to a likelihood function or not. I would guess yes such that it's a probability distribution. For consistency I think `neg_log_prior` or `negative_log_prior` would be the way to go.

Alternatively you can just update the docstrings to make this more obvious:
```
        likelihood
            The negative logarithm of the likelihood functional. Takes in the 
            observables and returns a form that can be assembled to 
            produce the value of the model-data misfit. Alternatively known 
            as the misfit functional.
        prior
            The negative log prior functional. Takes in the controls and 
            returns a form that can be assembled to produce the value of 
            the regularization. Alternatively known as the regularization 
            functional.
```

The only trouble here is that we are missing a term that can be `hidden physics`. Would that come under the likelihood (termed misfit here), the prior (termed regularization gere), or something else?

Reply:
This snippet of code is revealing

https://github.com/icepack/icepack/blob/242730cdcf7dfbfa6489da912fe3353c6784b1bb/test/statistics_test.py#L146-L159

here the likelihood really _isn't_ a likelihood - it's something that we want to minimise (the neg log likelihood as the docstring says). And yet the class that takes the `stats_problem` is called a `MaximumProbabilityEstimator`. This leads me to think that it really ought not be called `likelihood` either here or in the definition of `StatisticsProblem`

Reply:
Alright, I've changed it to `loss_functional` and `regularization`, the terminology is common enough in textbooks like Elements of Statistical Learning that people will get it.